### PR TITLE
express, mongodb integration

### PR DESCRIPTION
**Pull request by nikitahessde**

### CODE REVIEW

File: server/server.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import dotenv from 'dotenv'
+import express from 'express'
+import mongoose from 'mongoose'
+import taskRoutes from './router/tasks.js'
+import cors from 'cors'
```
Comment (Gints123) [inline server/server.js]: use typescript in BE aswell

**Review by Gints123 — CHANGES_REQUESTED** [review]

File: components/task-details.tsx [hunk 2]
```diff
@@ -4,9 +4,11 @@ import { useTasks } from "@/context/task-context";
 import { useState } from 'react';
 import ModeEdit from "@mui/icons-material/ModeEdit";
 import Check from "@mui/icons-material/Check";
+import { ObjectId } from "mongodb";
```
Comment (jevgenijs-jaunslavietis) [inline components/task-details.tsx]: Please don't use database models in UI

File: components/task-details.tsx [hunk 3]
```diff
@@ -93,14 +95,14 @@ export const TaskDetails = () => {
                                     task.status === 'inProgress' ? 'In Progress' : 
                                     'Done'}
                                 </td>
-                                <td className="py-3 px-6 text-left">{task.createdAt.toDateString()}</td>
+                                <td className="py-3 px-6 text-left">{new Date(task.createdAt).toDateString()}</td>
                                 <td className="py-3 px-6 text-left">
-                                    {editingTaskId === task.id ? (
-                                        <Check onClick={() => handleSave(task)} className="cursor-pointer">
+                                    {editingTaskId === task._id ? (
+                                        <Check onClick={() => handleSave(task as Task)} className="cursor-pointer">
```
Comment (jevgenijs-jaunslavietis) [inline components/task-details.tsx]: Something is not right with type. Normally, there should not be typecasting (`task as Task`)

File: context/task-context.tsx [hunk 4]
```diff
@@ -27,22 +29,59 @@ interface Task {
 export const TaskProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
     const [tasks, setTasks] = useState<Task[]>([]);
 
-    const addTask = (task: Task) => {
-        setTasks([...tasks, task]);
+    const fetchTasks = async () => {
+        const response = await fetch('http://localhost:4000/tasks');
```
Comment (jevgenijs-jaunslavietis) [inline context/task-context.tsx]: It would be better to extract base url into a separate const

File: package.json [hunk 5]
```diff
@@ -2,17 +2,24 @@
...
   "version": "0.1.0",
   "private": true,
+  "type": "module",
   "scripts": {
     "dev": "next dev",
     "build": "next build",
     "start": "next start",
-    "lint": "next lint"
+    "lint": "next lint",
+    "server": "node server/server.js"
```
Comment (jevgenijs-jaunslavietis) [inline package.json]: we don't need extra server in NextJs. Please check Next documentation

**Review by jevgenijs-jaunslavietis — CHANGES_REQUESTED** [review]

File: components/task-details.tsx [hunk 6]
```diff
@@ -4,9 +4,10 @@ import { useTasks } from "@/context/task-context";
 import { useState } from 'react';
 import ModeEdit from "@mui/icons-material/ModeEdit";
 import Check from "@mui/icons-material/Check";
+import { ObjectId } from "mongodb";
```
Comment (jevgenijs-jaunslavietis) [inline components/task-details.tsx]: I guess this import can be removed

File: components/task-list.tsx [hunk 7]
```diff
@@ -3,9 +3,10 @@
 import { DeleteOutline } from "@mui/icons-material";
 import { useTasks } from "@/context/task-context";
 import { Tooltip } from '@mui/material';
+import { ObjectId } from "mongodb";
```
Comment (jevgenijs-jaunslavietis) [inline components/task-list.tsx]: same as in previous comment

File: context/task-context.tsx [hunk 8]
```diff
@@ -1,13 +1,23 @@
 'use client';
 
-import { createContext, useContext, useState } from 'react';
+import { createContext, useContext, useEffect, useState } from 'react';
+import { ObjectId } from 'mongodb';
+
+interface Task {
+    _id?: ObjectId;
```
Comment (jevgenijs-jaunslavietis) [inline context/task-context.tsx]: same as before - DB models should not be in the UI

File: package.json (hunk 5 above)
Comment (jevgenijs-jaunslavietis) [inline package.json]: this is not needed as well

File: package.json [hunk 9]
```diff
@@ -2,17 +2,24 @@
...
   },
   "dependencies": {
     "@emotion/react": "^11.13.3",
     "@emotion/styled": "^11.13.0",
     "@mui/icons-material": "^6.1.4",
     "@mui/material": "^6.1.4",
+    "cors": "^2.8.5",
+    "dotenv": "^16.4.5",
+    "express": "^4.21.1",
```
Comment (jevgenijs-jaunslavietis) [inline package.json]: and this

File: package.json [hunk 10]
```diff
@@ -2,17 +2,24 @@
...
-    "lint": "next lint"
+    "lint": "next lint",
+    "server": "node server/server.js"
   },
   "dependencies": {
     "@emotion/react": "^11.13.3",
     "@emotion/styled": "^11.13.0",
     "@mui/icons-material": "^6.1.4",
     "@mui/material": "^6.1.4",
+    "cors": "^2.8.5",
```
Comment (jevgenijs-jaunslavietis) [inline package.json]: and this

File: tsconfig.json [hunk 11]
```diff
@@ -21,6 +21,6 @@
       "@/*": ["./*"]
     }
   },
-  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
+  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts", "app/api/tasks/[id]/route.ts"],
```
Comment (jevgenijs-jaunslavietis) [inline tsconfig.json]: is this change needed?

**Review by jevgenijs-jaunslavietis — CHANGES_REQUESTED** [review]

**Review by jevgenijs-jaunslavietis — APPROVED** [review]

File: app/api/tasks/[id]/route.ts
Comment (jevgenijs-jaunslavietis) [inline app/api/tasks/[id]/route.ts:1]: just FYI, we are usually not using routes in nextjs to execute server side code, but instead we use server actions for post/put/patch/delete and server components for get (e.g., in page.tsx).
